engine/src/main/java/io/zeebe/engine/processor/ProcessingStateMachine.java

```diff
@@ -359,8 +369,18 @@ private void writeEvent() {
             LOG.error(ERROR_MESSAGE_WRITE_EVENT_ABORTED, currentEvent, t);
             onError(t, this::writeEvent);
           } else {
-            updateState();
-            metrics.eventWritten();
+            actor.runOnCompletion(
+                future.get(),
+                (pos, t1) -> {
+                  if (t1 != null) {
+                    LOG.error(ERROR_MESSAGE_WRITE_EVENT_ABORTED, currentEvent, t1);
+                    onError(t1, this::writeEvent);
+                  } else {
+                    writtenEventPosition = pos;
+                    updateState();
+                    metrics.eventWritten();
+                  }
+                });
```

Comment:
@MiguelPires I think this is our problem why it is so slow - since we only continue with the processing when we have updated the records. I think we need to think about it if this is really necessary. Probably it is also fine to just continue with update state; Before we also just wrote it to the dispatcher - the writeEventPosition could be updated async.

Maybe you could try something like this?

```suggestion
 actor.runOnCompletion(
                future.get(),
                (pos, t1) -> {
                  if (t1 != null) {
                    LOG.error(ERROR_MESSAGE_WRITE_EVENT_ABORTED, currentEvent, t1);
                    onError(t1, this::writeEvent);
                  } else {
                    writtenEventPosition = pos;
                  }
                });
               updateState();
              metrics.eventWritten();
```